services/repository/push.go

```diff
@@ -0,0 +1,371 @@
+func (opts PushUpdateOptions) IsDelTag() bool {
+	return opts.IsTag() && opts.IsDelRef()
+}
+
+// IsBranch return true if it's a push to branch
+func (opts PushUpdateOptions) IsBranch() bool {
+	return strings.HasPrefix(opts.RefFullName, git.BranchPrefix)
+}
+
+// IsNewBranch return true if it's the first-time push to a branch
+func (opts PushUpdateOptions) IsNewBranch() bool {
+	return opts.IsBranch() && opts.IsNewRef()
+}
+
+// IsUpdateBranch return true if it's not the first push to a branch
+func (opts PushUpdateOptions) IsUpdateBranch() bool {
+	return opts.IsBranch() && opts.IsUpdateRef()
+}
+
+// IsDelBranch return true if it's a deletion to a branch
+func (opts PushUpdateOptions) IsDelBranch() bool {
+	return opts.IsBranch() && opts.IsDelRef()
+}
+
+// TagName returns simple tag name if it's an operation to a tag
+func (opts PushUpdateOptions) TagName() string {
+	return opts.RefFullName[len(git.TagPrefix):]
+}
+
+// BranchName returns simple branch name if it's an operation to branch
+func (opts PushUpdateOptions) BranchName() string {
+	return opts.RefFullName[len(git.BranchPrefix):]
+}
+
+// RepoFullName returns repo full name
+func (opts PushUpdateOptions) RepoFullName() string {
+	return opts.RepoUserName + "/" + opts.RepoName
+}
+
+// pushQueue represents a queue to handle update pull request tests
+var pushQueue queue.Queue
+
+// handle passed PR IDs and test the PRs
+func handle(data ...queue.Data) {
+	for _, datum := range data {
+		opts := datum.([]*PushUpdateOptions)
+		if err := pushUpdates(opts); err != nil {
+			log.Error("pushUpdate failed: %v", err)
+		}
+	}
+}
+
+func initPushQueue() error {
+	pushQueue = queue.CreateQueue("push_update", handle, []*PushUpdateOptions{}).(queue.Queue)
+	if pushQueue == nil {
+		return fmt.Errorf("Unable to create push_update Queue")
+	}
+
+	go graceful.GetManager().RunWithShutdownFns(pushQueue.Run)
+	return nil
+}
+
+// PushUpdate is an alias of PushUpdates for single push update options
+func PushUpdate(opts *PushUpdateOptions) error {
+	return PushUpdates([]*PushUpdateOptions{opts})
+}
+
+// PushUpdates adds a push update to push queue
+func PushUpdates(opts []*PushUpdateOptions) error {
+	if len(opts) == 0 {
+		return nil
+	}
+
+	for _, opt := range opts {
+		if opt.IsNewRef() && opt.IsDelRef() {
+			return fmt.Errorf("Old and new revisions are both %s", git.EmptySHA)
+		}
+	}
+
+	return pushQueue.Push(opts)
+}
+
+// pushUpdates generates push action history feeds for push updating multiple refs
+func pushUpdates(optsList []*PushUpdateOptions) error {
+	if len(optsList) == 0 {
+		return nil
+	}
+
+	repo, err := models.GetRepositoryByOwnerAndName(optsList[0].RepoUserName, optsList[0].RepoName)
+	if err != nil {
+		return fmt.Errorf("GetRepositoryByOwnerAndName failed: %v", err)
+	}
+
+	repoPath := repo.RepoPath()
+	_, err = git.NewCommand("update-server-info").RunInDir(repoPath)
+	if err != nil {
+		return fmt.Errorf("Failed to call 'git update-server-info': %v", err)
+	}
```

Comment:
I'm concerned that this is outdated - update-server-info is only used for dumb HTTP and therefore this should be removed.

It's also running at the wrong time here. If we want to use update-server-info we should be using a git post-update hook.

Reply:
then again it appears we do support the `info/refs` endpoint ... hmm ...

Reply:
hmm old code has it too - so I'm ok with it - if this gets problematic I think this can have its own pull

👇 so LGTM from my side :)